app.js

```diff
@@ -20,154 +20,154 @@ const swaggerUi = require('swagger-ui-express')
-  search.initialize()
-  if (searchProvider === 'memory') definitionService.reload('definitions')
-})
 
-const definitionService = require('./business/definitionService')(
-  harvestStore,
-  summaryService,
-  aggregatorService,
-  curationService,
-  definitionStore,
-  search
-)
-// Circular dependency. Reach in and set the curationService's definitionService. Sigh.
-curationService.definitionService = definitionService
-const definitions = require('./routes/definitions')(definitionService)
-
-const attachments = require('./routes/attachments')(harvestStore)
-
-const appLogger = console // @todo add real logger
-const githubSecret = config.webhook.githubSecret
-const crawlerSecret = config.webhook.crawlerSecret
-const webhook = require('./routes/webhook')(curationService, definitionService, appLogger, githubSecret, crawlerSecret)
-
-const cachingProvider = config.caching.provider
-const caching = require(`./providers/caching/${cachingProvider}`)
-const cachingMiddleware = require('./middleware/caching')
-
-const app = express()
-app.use(cors())
-app.options('*', cors())
-app.use(cookieParser())
-app.use(helmet())
-app.use(requestId())
-app.use(cachingMiddleware(caching()))
-
-app.use(logger('dev'))
-app.use(configMiddleware)
-app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDoc))
-app.use('/webhook', bodyParser.raw({ limit: '5mb', type: '*/*' }), webhook)
-
-// OAuth app initialization; skip if not configured (middleware can cope)
-if (config.auth.github.clientId) {
-  passport.use(auth.getStrategy())
-  app.use(passport.initialize())
-}
-app.use('/auth', auth())
-app.use(githubMiddleware)
-
-// rate-limit the remaining routes
-app.set('trust-proxy', true)
-app.use(
-  new RateLimit({
-    windowMs: config.limits.windowSeconds * 1000,
-    max: config.limits.max,
-    delayAfter: 0
+function createApp(config) {
+  const initializers = []
+  const summaryService = require('./business/summarizer')(config.summary)
+
+  const harvestStore = config.harvest.store()
+  initializers.push(async () => harvestStore.initialize())
+  const harvestService = config.harvest.service()
+  const harvest = require('./routes/harvest')(harvestService, harvestStore, summaryService)
+
+  const aggregatorService = require('./business/aggregator')(config.aggregator)
+
+  const curationService = config.curation.service(null, config.endpoints)
+  const curations = require('./routes/curations')(curationService)
+
+  const definitionStore = config.definition.store()
+  initializers.push(async () => definitionStore.initialize())
+
+  const attachmentStore = config.attachment.store()
+  initializers.push(async () => attachmentStore.initialize())
+
+  const searchService = config.search.service()
+  initializers.push(async () => {
+    await searchService.initialize()
+    // Bit of trick for local hosting. Preload search if using an in-memory search service
+    if (searchService.constructor.name === 'MemorySearch') definitionService.reload('definitions')
```

Comment:
what do you think about moving this into the requestHandler.init right after?

```js
console.log('Service initialized')
```

I was getting a bunch of calls to `store()` in mongo provider before the collection was initialized

Reply:
Yeah, there is a race between the mongo store being initialized and the this search reloading. Those are both in the same set of initializers. They are run in parallel so it is not guaranteed that the definition store is initialized before this is run. This works in production as the search service is not 'memory' so this is not done. Even still, seems like a good move.

On a related note, when does the mongo collection get created. in mongo.js `initialize` we have code to create the db and the collection but it is unclear if those are local client-only calls or if they talk to the service and need to be awaited or something. Do we need to explicitly create the collection like we do with the container in blob?

Reply:
so are you getting errors from trying to read before the collection is created? That is a different problem. We are not proactively creating the collection so until we upsert it may not be there. I'm hoping we don't have to do a bogus upsert on every start just to keep `find` happy.

Or are you saying that it is failing before `this.collection` the property is set?

Reply:
`this.collection` is undefined when the reload happens.

We shouldn't need to worry about bogus upsert or anything I think the initializers just need to finish